scripts/composer/ScriptHandler.php

```diff
@@ -88,6 +88,77 @@ public static function downloadDevelopPackages(Event $event) {
     }
   }
 
+  /**
+   * Reset local repositories to the default branch.
+   *
+   * @param \Composer\Script\Event $event
+   *   The script event.
+   */
+  public static function resetLocalRepositories(Event $event) {
+    $io = $event->getIO();
+    $repositoriesInfo = self::getLocalRepositoriesInfo($event);
+    $drupalFinder = new DrupalFinder();
+    $drupalFinder->locateRoot(getcwd());
+    $composerRoot = $drupalFinder->getComposerRoot();
+
+    $io->write(PHP_EOL);
+    foreach ($repositoriesInfo as $key => $info) {
+      $localBranch = trim(shell_exec('git -C ' . $composerRoot . '/' . $info['install_path'] . ' rev-parse --abbrev-ref HEAD'));
+
+      if ($localBranch === $info['branch']) {
+        $io->write($info['package'] . ' is already on branch ' . $info['branch'], TRUE);
+        continue;
+      }
+
+      $gitStatus = shell_exec('git -C ' . $composerRoot . '/' . $info['install_path'] . ' status --porcelain --untracked-files=no');
```

Comment:
I would also include untracked files, because it could be that you add new files that could make effect to functionality of other branch (fe. *.module file with hooks or similar), that's why it would be good to get rid of that on clean checkout.

And then we can use `[-u|--include-untracked]` for `git stash`.